pageserver/src/tenant/mgr.rs

```diff
@@ -1350,6 +1350,9 @@ impl TenantManager {
         }
     }
 
+    /// Deletes the tenant shard data from remote storage. If an unsharded tenant ID is given for a
+    /// sharded tenant, it will remove all shards for that tenant (i.e. all paths with the tenant ID
+    /// prefix) -- including stale shards.
     async fn delete_tenant_remote(
         &self,
         tenant_shard_id: TenantShardId,
```

Comment:
I can't think of a case when we would ever want to delete the remote data for just one shard. Should this function take a `TenantId` as its argument instead? Allows one to skip reading the comment.

Reply:
Yeah, I also found it a bit iffy, but figured it was like this for a reason -- maybe for shard cleanup or something. I'll look around to see if it'd break anything.

Reply:
Had a look at the pageservers logs, and didn't see any tenant deletion calls that specified a shard ID in the past 7 days, so this should be ok. Also verified that the control plane only deletes complete tenants:

https://github.com/neondatabase/cloud/blob/3346ae102db10c05988cb7933368800e65fa2acd/goapp/controlplane/internal/pkg/pageservercontroller/controller.go#L485-L489

I'll submit the API change as a separate PR.

Reply:
On second thought, I don't think it makes sense for the Pageserver's `delete_tenant()` to be scoped to a `TenantId` rather than a `TenantShardId`. Afaict the Pageserver generally operates on shards rather than entire tenants (it kind of has to since it has an incomplete set of shards) -- this includes stuff like internal locks (e.g. `TenantsMap`). All other API methods appear to be shard-scoped as well.

The impedance mismatch here really comes from wanting to run a tenant-scoped operation via a shard-scoped API. I think ideally we'd want to lift the tenant-wide operation up into the storage controller, and leave pageserver operations shard-scoped -- but it doesn't seem like the storage controller currently has infrastructure to interact with remote storage. Is that worth considering?

Barring that, it might be better to add a separate pageserver API method to explicitly delete all remote data from a tenant, rather than overloading the shard deletion API. I suppose the shard deletion API could be removed entirely once all deletions go via the storage controller (since it always detaches first). We could also just leave this cleanup to the storage scrubber, and perhaps ask it to eagerly scrub the tenant on deletion.

Reply:
> The impedance mismatch here really comes from wanting to run a tenant-scoped operation via a shard-scoped API. I think ideally we'd want to lift the tenant-wide operation up into the storage controller, and leave pageserver operations shard-scoped -- but it doesn't seem like the storage controller currently has infrastructure to interact with remote storage. Is that worth considering?

Agree with this point. Deletions can run from any box as far as I understand (lambda, storcon, etc.)

Reply:
Ok, let me try lifting it up into storcon.

Reply:
Following discussion in #9394 and [Slack](https://neondb.slack.com/archives/C033RQ5SPDH/p1728993338890879), we'll keep the special case in the pageserver delete tenant API for now. It would be nice to move this logic somewhere else, but it's not clear where it belongs (possibly a separate service for background operations and housekeeping).